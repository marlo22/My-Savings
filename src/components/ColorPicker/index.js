import React, { useState } from 'react';
import styled from 'styled-components/native';
import PropTypes from 'prop-types';

import { Button } from 'native-base';
import { TriangleColorPicker, fromHsv } from 'react-native-color-picker';

import { primaryContrastColor, borderRadius } from '../../themes';

const ModalWrapper = styled.View`
  background: rgba(0, 0, 0, 0.7);
  width: 100%;
  height: 100%;
  justify-content: center;
  align-items: center;
  position: absolute;
`;

const StyledModal = styled.View`
  width: 90%;
  height: 60%;
  background-color: #ffffff;
  border-radius: ${borderRadius};
`;

const ButtonsGroup = styled.View`
  flex-direction: row;
  justify-content: flex-end;
  margin: 10px auto;
`;

const FooterButton = styled(Button)`
  margin: 0 5px;
`;

const ButtonText = styled.Text`
  color: ${primaryContrastColor};
`;

const ColorPicker = ({
  onOk,
  onClose,
  pickerProps = {}
}) => {
  const [value, setValue] = useState('');

  const handleSave = () => {
    const { h, s, v } = value;
    const hexValue = fromHsv({ h, s, v });

    onOk(hexValue);
    onClose();
  }

  return (
    <ModalWrapper visible={true} elevation={999}>
      <StyledModal>
        <TriangleColorPicker
          style={{ flex: 1 }}
          onColorChange={setValue}
          {...pickerProps}
        />
        <ButtonsGroup>
          <FooterButton onPress={onClose}>
            <ButtonText>Anuluj</ButtonText>
          </FooterButton>
          <FooterButton onPress={handleSave}>
            <ButtonText>Wybierz</ButtonText>
          </FooterButton>
        </ButtonsGroup>
      </StyledModal>
    </ModalWrapper>
  );
};

ColorPicker.propTypes = {
  onOk: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  pickerProps: PropTypes.object
};

export default ColorPicker;