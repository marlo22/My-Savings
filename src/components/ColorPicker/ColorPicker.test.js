import React from 'react';
import { shallow } from 'enzyme';

import ColorPicker from './';

const buttonDisplayName = 'Styled(Styled(Button))';

const onOk = jest.fn();
const onClose = jest.fn();
const pickerProps = { defaultColor: '#ff0000' };

const component = shallow(
  <ColorPicker
    onOk={onOk}
    onClose={onClose}
    pickerProps={pickerProps}
  />
);

describe('ColorPicker', () => {
  test('should render without errors', () => {
    expect(component).toBeTruthy();
  });

  test('snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  test('should set passed picker props', () => {
    const picker = component.find('TriangleColorPicker');
    expect(picker.prop('defaultColor')).toBe(pickerProps.defaultColor);
  });

  test('pickerColorChange should change value', () => {
    const hsvValue = { h: 134, s: 14, v: 35 };
    const hexValue = '#4d5950';

    const picker = component.find('TriangleColorPicker');
    picker.props().onColorChange(hsvValue);
    
    const saveButton = component.find(buttonDisplayName).last();
    saveButton.simulate('press');
    expect(onOk).toHaveBeenCalledWith(hexValue);
  });

  describe('buttons', () => {
    afterEach(() => {
      onOk.mockClear();
      onClose.mockClear();
    });

    test('click first button should call onClose function', () => {
      const button = component.find(buttonDisplayName).first();
      
      button.simulate('press');
      expect(onClose).toHaveBeenCalled();
    });

    test('click last button should call onOk and onClose function', () => {
      const button = component.find(buttonDisplayName).last();
      
      button.simulate('press');
      expect(onOk).toHaveBeenCalled();
      expect(onClose).toHaveBeenCalled();
    });
  });
});