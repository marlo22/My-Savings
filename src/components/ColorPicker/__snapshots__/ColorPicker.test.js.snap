// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker snapshot 1`] = `
<Styled(View)
  elevation={999}
  visible={true}
>
  <Styled(View)>
    <TriangleColorPicker
      defaultColor="#ff0000"
      onColorChange={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Styled(View)>
      <Styled(Styled(Button))
        onPress={[MockFunction]}
      >
        <Styled(Text)>
          Anuluj
        </Styled(Text)>
      </Styled(Styled(Button))>
      <Styled(Styled(Button))
        onPress={[Function]}
      >
        <Styled(Text)>
          Wybierz
        </Styled(Text)>
      </Styled(Styled(Button))>
    </Styled(View)>
  </Styled(View)>
</Styled(View)>
`;
